scripts/prepare-test-ledger.sh

```diff
@@ -1,86 +1,158 @@
+# Download ledger
+##########################################################
 
-ledger_file="$epoch_now.json"
+ledger_file="$LEDGER_PREFIX.json"
 
-echo "Current epoch: $epoch_now" >&2
+echo "Using ledger with prefix: $LEDGER_PREFIX" >&2
 
 if [[ ! -f "$ledger_file" ]]; then
-  ledgers_url="https://storage.googleapis.com/storage/v1/b/mina-staking-ledgers/o?maxResults=1&prefix=staking-$epoch_now"
-  ledger_url=$(curl "$ledgers_url" | jq -r .items[0].mediaLink)
-
-  wget -O $ledger_file "$ledger_url"
+  ledgers_url="https://storage.googleapis.com/storage/v1/b/mina-staking-ledgers/o?maxResults=1000&prefix=$LEDGER_PREFIX"
+  echo "$ledgers_url" >&2
+  ledger_url=$(curl "$ledgers_url" | jq -r '.items | sort_by(.size|tonumber) | last.mediaLink')
+  if [[ "$ledger_url" == null ]]; then
+    echo "Couldn't fine ledger with prefix $LEDGER_PREFIX" >&2
+    exit 2
+  fi
+  curl "$ledger_url" >"$ledger_file"
+  cmp "$ledger_file" >/dev/null <<< "Ledger not found: next staking ledger is not finalized yet" && echo "Next ledger not finalized yet" >&2 && rm "$ledger_file" && exit 2
 fi
 
 keys_=""
-for key in "$@"; do
+for key in "${KEYS[@]}"; do
   keys_="\"$key\",$keys_"
 done
 keys_="${keys_:0:-1}"
 
-# jq filter to exclude block PKs from the ledger
-pre_filter="[.[] | select(.pk | IN($keys_) | not)]"
+tmpfile=$(mktemp)
+
+# jq filter to exclude PKs from the ledger
+<"$ledger_file" jq "[.[] | select(.pk | IN($keys_) | not)]" >"$tmpfile"
+
+num_accounts=$(<"$tmpfile" jq length)
 
 ##########################################################
-# Calculate and print approximate new balances
+# Create new ledger in a temporary file
 ##########################################################
 
-num_accounts=$(<$ledger_file jq "$pre_filter | length")
-total_balance=$(<$ledger_file jq "$pre_filter | [.[].balance | tonumber] | add | round")
-new_total_balance=$((total_balance+num_keys*KEY_BALANCE))
-echo "Total accounts: $num_accounts, balance: $total_balance, new balance: $new_total_balance" >&2
+TOP_KEYS=()
+if [[ "$REPLACE_TOP" != "" ]]; then
+  top_expr="[group_by(.delegate)[] | [(map(.balance|tonumber)|add), .[0].delegate]] | sort | reverse | map(.[1]) | .[0:$num_keys][]"
+  readarray -t TOP_KEYS < <( jq -r "$top_expr" "$tmpfile" )
+fi
 
-function make_balance_expr(){
+function make_expr(){
   i=$1
-  key=$2
-  echo "$key: (((([.[range($i; $num_accounts; $num_keys_and_ghost)].balance | tonumber] | add) + $KEY_BALANCE)/$new_total_balance*10000|round/100 | tostring) + \"%\")"
+  key="${KEYS[$i]}"
+
+  if [[ "$REPLACE_TOP" == "" ]]; then
+    interval=$num_keys
+    if [[ "$NORM" != "" ]]; then
+      interval=$(( (RANDOM+RANDOM+RANDOM)*num_keys/32/1024/3+1 ))
+    fi
+    # Substitute delegate of some ledger keys with the BP key
+    echo "((.[range($i; $num_accounts; $interval)] | select ((.balance|tonumber) > $DELEGATEE_CUTOFF)).delegate |= \"$key\")"
+  else
+    echo "((.[] | select(.delegate == \"${TOP_KEYS[$i]}\")).delegate |= \"$key\")"
```

Comment:
this is a lot cleaner than i thought it was going to be, nice.

Reply:
I have a version that's even cleaner (and twice shorter in code), however for some reason it's twice slower in terms of performance (and this version takes ~15 minutes on my laptop, so it matters).

Surprisingly, replacing keys is much more expensive than randomly overwriting them based on index ranges.